LambdaEngine/Source/ECS/ECSCore.cpp

```diff
@@ -2,128 +2,155 @@
-
-        for (uint32 entityIdx = 0; entityIdx < entities.GetSize(); entityIdx++)
-        {
-            const std::unordered_set<std::type_index>& typeSet = entityComponentSets[entityIdx];
-
-            for (std::type_index type : typeSet) {
-                // Deregister entity's components from systems
-                m_EntityPublisher.UnpublishComponent(entities[entityIdx], type);
-            }
-        }
-    }
-
-    void ECSCore::DeleteTopRegistryPage()
-    {
-        const EntityRegistryPage& page = m_EntityRegistry.GetTopRegistryPage();
-        const auto& entityComponentSets = page.GetVec();
-        const TArray<Entity>& entities = page.GetIDs();
-
-        for (uint32 entityIdx = 0; entityIdx < entities.GetSize(); entityIdx++)
-        {
-            Entity entity = entities[entityIdx];
-            const std::unordered_set<std::type_index>& typeSet = entityComponentSets[entityIdx];
-
-            for (std::type_index componentType : typeSet)
-            {
-                // Deregister entity's component from systems
-                m_EntityPublisher.UnpublishComponent(entities[entityIdx], componentType);
-
-                // Delete the component
-                m_ComponentStorage.GetComponentArray(componentType)->Remove(entity);
-            }
-        }
-
-        m_EntityRegistry.RemovePage();
-    }
-
-    void ECSCore::ReinstateTopRegistryPage()
-    {
-        const EntityRegistryPage& page = m_EntityRegistry.GetTopRegistryPage();
-
-        const auto& entityComponentSets = page.GetVec();
-        const TArray<Entity>& entities = page.GetIDs();
-
-        for (uint32 entityIdx = 0; entityIdx < entities.GetSize(); entityIdx++)
-        {
-            const std::unordered_set<std::type_index>& typeSet = entityComponentSets[entityIdx];
-
-            for (std::type_index componentType : typeSet)
-            {
-                m_EntityPublisher.PublishComponent(entities[entityIdx], componentType);
-            }
-        }
-    }
-
-    void ECSCore::PerformEntityDeletions()
-    {
-        const EntityRegistryPage& registryPage = m_EntityRegistry.GetTopRegistryPage();
-
-        for (Entity entity : m_EntitiesToDelete)
-        {
-            // Delete every component belonging to the entity
-            const auto& componentTypes = registryPage.IndexID(entity);
-            for (std::type_index componentType : componentTypes)
-            {
-                // Delete the component
-                IComponentArray* pComponentArray = m_ComponentStorage.GetComponentArray(componentType);
-                pComponentArray->Remove(entity);
-
-                // Notify systems that the component has been removed
-                m_EntityPublisher.UnpublishComponent(entity, componentType);
-            }
-
-            // Free the entity ID
-            m_EntityRegistry.DeregisterEntity(entity);
-        }
-
-        m_EntitiesToDelete.Clear();
-    }
+	ECSCore* ECSCore::s_pInstance = DBG_NEW ECSCore();
```

Comment:
Is it necessary to dynamically allocate the ECSCore Instance? Could it not be statically initialized instead?

Reply:
When systems are deleted, they call `ECSCore::UnsubscribeFromEntities`. Since both `ECSCore` and some systems are static, there is no guarantee that systems will be deleted before `ECSCore`.

By making the instance a pointer, systems can check if `ECSCore` is `nullptr` or not, before attempting to call its member function.